src/v/ssx/metrics.h

```diff
@@ -81,18 +83,105 @@ class metric_groups {
     // Add metrics belonging to the same group.
     //
     // This is the same method as above but using an explicit vector.
-    metric_groups& add_group(
+    metric_groups_base& add_group(
       const ss::metrics::group_name_type& name,
       const std::vector<ss::metrics::metric_definition>& l) {
         _underlying.add_group(name, l);
         return *this;
     }
 
-    // Clear all metric registrations.
-    void clear() { _underlying.clear(); }
+protected:
+    ss::metrics::metric_groups _underlying;
+};
+
+class internal_metric_groups : public metric_groups_base {
+public:
+    internal_metric_groups()
+      : metric_groups_base(ss::metrics::default_handle()) {}
+
+    using metric_groups_base::add_group;
+
+    /*
+     * The two following methods are an extension to the methods above. They
+     * allow registering the metrics in the same way. However when registering
+     * metrics with label aggregation prefer to use this method.
+     *
+     * It allows specifying the aggregation labels to use when the config flag
+     * `aggregate_metrics=false` as well as `aggregate_metrics=true`.
+     *
+     * Further it uses the `metrics_registry` under the hood such that the
+     * aggregation labels will be automatically updated when the config flag
+     * changes.
+     *
+     * The specified labels will be used for all the specified metrics in the
+     * group. If aggregation labels are specified on a per metric basis they
+     * will be overriden.
+     *
+     * Implementation note: Taking `impl::metric_definition_impl` isn't great
+     * but that's a consequence of how the seastar metrics API works.
+     * `ss::metrics::metric_definition` is just a thin wrapper around
+     * `impl::metric_definition_impl` which doesn't expose any of the members
+     * (why?). Further it actually takes a `impl::metric_definition_impl` as its
+     * main constructor argument which is returned by all the `make_` functions.
+     * Hence we are forced to use the `impl` type here as well as we need access
+     * to the metric names.
+     */
+    internal_metric_groups& add_group(
```

Comment:
This is a great comment, but as it will be a heavily used method well into the future by lots of people with only a passing interest in metrics, can we carefully document this function in a "usage" sense, like mentioning what the function does, what all of the parameters mean, etc.

The existing doc is great, you could keep it here or move it to the class doc or inside the function as an impl doc (which would keep the "usage" doc which would pop up in everyone's editor succinct).

Reply:
Made it proper doxygen style and added an example. Moved parts of it into the function.

Reply:
Looks great, thanks!